### PR TITLE
PIMS 1957/1958: Handle agency response change notifications

### PR DESCRIPTION
**Pull request by GrahamS-Quartech**
## 🎯 Summary

  
[PIMS-1957](https://apps.itsm.gov.bc.ca/jira/browse/PIMS-1957)
[PIMS-1958](https://apps.itsm.gov.bc.ca/jira/browse/PIMS-1958)

* Added a helper function to projectServices which will compare old agency response to new responses and check for differences. 
* Added a helper function to utils to get the number of days elapsed between two UTC dates.
* Adds a new function, generateProjectWatchNotifications, which mirrors the functionality of a similarly named function in the old backend. This function will cancel notifications if an agency response was moved to a non-subscribe status, or queue new notifications if an agency was newly added to subscribe.
* Modified the handleProjectNotifications function to trigger both generateProjectNotifications and generateProjectWatchNotifications. 

## 🔰 Checklist

- [x] I have read and agree with the following checklist and am following the guidelines in our [Code of Conduct](CODE_OF_CONDUCT.md) document.

> - I have performed a self-review of my code.
> - I have commented my code, particularly in hard-to-understand areas.
> - I have made corresponding changes to the documentation where required.
> - I have tested my changes to the best of my ability.
> - My changes generate no new warnings.

### CODE REVIEW

File: express-api/src/services/projects/projectsServices.ts [hunk 1]
```diff
@@ -601,6 +625,31 @@ const handleProjectMonetary = async (
...
+const getAgencyResponseChanges = async (
+  oldProject: Project,
+  newProject: DeepPartial<Project>,
+): Promise<Array<ProjectAgencyResponse>> => {
+  const retResponses: Array<ProjectAgencyResponse> = [];
+  for (const response of newProject.AgencyResponses as ProjectAgencyResponse[]) {
+    const originalResponse = oldProject.AgencyResponses.find(
+      (r) => r.AgencyId === response.AgencyId,
+    );
+    if (originalResponse == null || originalResponse.Response != response.Response) {
```
[the remark this answers is not in the document]
Comment (GrahamS-Quartech) [reply express-api/src/services/projects/projectsServices.ts:637]: I had this same thought when looking at the old backend for this, but it's actually the field "Response" of an agency response object being compared here, so the comparison is between numbers.

**Comment by LawrenceLau2020** [comment]
Testing the "unsubscribe" options seems to work fine, after choosing an agency that had already had pending notifications, I added this agency and after clicking confirm I needed to refresh the page in order to see the updated table with the 3 cancelled notifications. Perhaps we can refresh the notifications table after the "confirm"?

When testing the "subscribe" option, I was expecting to receive an email as well as seeing the notifications table get updated with the new agency interest, but didn't get an email so maybe this isn't being triggered after clicking "confirm"?

image

**Comment by LawrenceLau2020** [comment]
Re-tested the agency interest by adding an interest for Emily Carr University, and there were 3 new pending notifications in the table after confirming. Then changed the interest to "Watch" which cancels the notification, and then changing the interest back to "Subscribe" will create 3 new "pending" notifications for this agency as well. 

The notifications table is refreshing nicely as well, so I'm going to approve this ticket.

image

**Review by LawrenceLau2020 — APPROVED** [review]

**Review by dbarkowsky — APPROVED** [review]
Last changes made the emails scheduled to send at correct times!

**Pull request merged**